mjx/mujoco/mjx/_src/sensor.py

```diff
@@ -600,3 +602,79 @@ def _framelinacc(cvel, cacc, offset):
   )
 
   return d.replace(sensordata=sensordata)
+
+
+def energy_pos(m: Model, d: Data) -> Data:
+    """Calculates position-dependent energy (potential).
+    """
+        
+    # Initialize potential energy
+    energy = jp.array(0.0)
+    if not m.opt.enableflags & EnableBit.ENERGY:
+        return d
+    
+    # Add gravitational potential energy for each body
+    if not m.opt.disableflags & DisableBit.GRAVITY:
+        for i in range(1, m.nbody):  # Skip world body
+            energy -= m.body_mass[i] * jp.dot(m.opt.gravity, d.xipos[i])
+    
+    # Add joint spring potential energy
+    if not m.opt.disableflags & DisableBit.PASSIVE:
+        for i in range(m.njnt):
+            stiffness = m.jnt_stiffness[i]
+            padr = m.jnt_qposadr[i]
+            
+            if m.jnt_type[i] == JointType.FREE:
+                # Position springs
+                quat = d.qpos[padr:padr+4]
+                quat = math.normalize(quat)
+                dif = quat - m.qpos_spring[padr:padr+4]
+                energy += 0.5 * stiffness * jp.dot(dif[:3], dif[:3])
+                
+                # Handle rotations
+                padr += 3
+                
+            if m.jnt_type[i] in (JointType.FREE, JointType.BALL):
+                # Convert quaternion difference to angular displacement
+                quat = d.qpos[padr:padr+4]
+                quat = math.normalize(quat)
+                dif = math.quat_sub(quat, m.qpos_spring[padr:padr+4])
+                energy += 0.5 * stiffness * jp.dot(dif, dif)
+                
+            elif m.jnt_type[i] in (JointType.SLIDE, JointType.HINGE):
+                dif = d.qpos[padr] - m.qpos_spring[padr]
+                energy += 0.5 * stiffness * dif * dif
+
+    # Add tendon spring potential energy
+    if not m.opt.disableflags & DisableBit.PASSIVE:
+        for i in range(m.ntendon):
```

Comment:
this for loop can be removed, and you should not need `scan` at all.

you can do something like:

```py
lower = m.tendon_lengthspring[0::2]
upper = m.tendon_lengthspring[1::2]
length = d.ten_length

displacement = jp.where(length > upper, upper - length, 0)
displacement = jp.where(length < lower, lower - length, displacement)
...
```